src/Compiler/Service/TransparentCompiler.fs

```diff
@@ -384,10 +384,11 @@ type internal TransparentCompiler
         (useSdkRefs: bool option)
         (sdkDirOverride: string option)
         (assumeDotNetFramework: bool option)
-        (projectSnapshot: ProjectSnapshot)
+        (fileKey: ICacheKey<string * ProjectIdentifier, string>)
```

Comment:
This is looking a little bit suspicious.

Can we instead generate the appropriate key inside this function?

Reply:
We can keep the projectSnapshot if it is ok to create a temporary one for the script call.

Reply:
It's fine, but maybe it's not necessary? It seems we only need it for `OtherOptions` which could be passed separately. We can just create a custom `ICacheKey` in that function that accurately reflects what's necessary to compute a valid value for the input. Theoretically it could even save us some re-computations when something in snapshot is different, but it won't actually affect the script closure.

Reply:
Yeah we can add that also as input. But we probably don't need all the other files and project dependencies that are also contained in the snapshot.

Reply:
You would compose a new key inside that function based on the sourcetext and other options?
Any pointers on how I can easily do that?

Reply:
Well it has a bunch of other inputs too that will presumably affect the result. So actually I don't know if even the snapshot is sufficient for all that. Maybe it is if all the other inputs we give it are also previously computed using that same snapshot. But we're calling it from multiple places now so that's a bit tough to think about. 

So I'd propose to create a new `ICacheKey` where,

`Key` = something that uniquely identifies the job - for which we only want to keep one result at a time. So probably something like file name + some of those boolean flags. Maybe ProjectIdentifier. I don't really understand the script closures and how they're used well enough to tell. But basically this is used to discard old data when there's a new version of it - but if it's not specific enough (let's say just file name) the multiple places where we use it could evict each other's results when they're still valid.

`Version` = everything that affects the result, so sourceText (checksum or hash), other options, all the switches. You can use the `MD5Hasher` module to put a bunch of things in a hash, or it can just be a tuple. 

Hope this makes some sense. Or just ping me on slack and we can discuss it more.

Reply:
Thanks, I think I understand what we are trying to do.
I made something, but I'm aware probably still a little off.